File: scripts/optimize-document-assets.mjs
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { createRequire } from "module";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const publicDir = path.join(__dirname, "../client/public");
const require = createRequire(import.meta.url);

async function optimize() {
  let sharp;
  try {
    sharp = require("sharp");
  } catch {
    console.log("sharp not installed, using built-in fallback via child process...");
    const { execSync } = await import("child_process");
    execSync(`python "${path.join(__dirname, "optimize_document_assets.py")}"`, { stdio: "inherit" });
    return;
  }

  const assets = [
    {
      input: "document-assets/solar-document-header.png",
      outputs: [{ file: "document-assets/solar-document-header.jpg", width: 1200, quality: 82 }],
    },
    {
      input: "document-assets/vendor-authorized-signature.png",
      outputs: [{ file: "document-assets/vendor-authorized-signature.png", width: 480, png: true }],
    },
    {
      input: "logo.png",
      outputs: [{ file: "logo.jpg", width: 512, quality: 85 }],
    },
    {
      input: "document-assets/agreement-stamp-paper.png",
      outputs: [{ file: "document-assets/agreement-stamp-paper.jpg", width: 800, quality: 85 }],
    },
  ];

  for (const asset of assets) {
    const inputPath = path.join(publicDir, asset.input);
    if (!fs.existsSync(inputPath)) {
      console.warn("Skip missing:", asset.input);
      continue;
    }
    for (const out of asset.outputs) {
      const outputPath = path.join(publicDir, out.file);
      let pipeline = sharp(inputPath).rotate().resize({ width: out.width, withoutEnlargement: true });
      if (out.png) {
        await pipeline.png({ compressionLevel: 9, palette: true }).toFile(outputPath);
      } else {
        await pipeline.jpeg({ quality: out.quality, mozjpeg: true }).toFile(outputPath);
      }
      const kb = Math.round(fs.statSync(outputPath).size / 1024);
      console.log(`Optimized ${out.file} (${kb} KB)`);
    }
  }
}

optimize().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/optimize_document_assets.py
from PIL import Image
import os

ROOT = os.path.join(os.path.dirname(__file__), "..", "client", "public")

def save_jpg(src_rel, out_rel, max_w, quality=82):
    src = os.path.join(ROOT, src_rel)
    out = os.path.join(ROOT, out_rel)
    img = Image.open(src)
    if img.mode in ("RGBA", "P"):
        bg = Image.new("RGB", img.size, (255, 255, 255))
        bg.paste(img, mask=img.split()[-1] if img.mode == "RGBA" else None)
        img = bg
    else:
        img = img.convert("RGB")
    w, h = img.size
    if w > max_w:
        nh = int(h * max_w / w)
        img = img.resize((max_w, nh), Image.LANCZOS)
    img.save(out, "JPEG", quality=quality, optimize=True, progressive=True)
    print(f"Optimized {out_rel} ({os.path.getsize(out)//1024} KB)")

def save_png(src_rel, out_rel, max_w):
    src = os.path.join(ROOT, src_rel)
    out = os.path.join(ROOT, out_rel)
    img = Image.open(src)
    w, h = img.size
    if w > max_w:
        nh = int(h * max_w / w)
        img = img.resize((max_w, nh), Image.LANCZOS)
    img.save(out, "PNG", optimize=True)
    print(f"Optimized {out_rel} ({os.path.getsize(out)//1024} KB)")

if __name__ == "__main__":
    save_jpg("document-assets/solar-document-header.png", "document-assets/solar-document-header.jpg", 1200, 82)
    save_png("document-assets/vendor-authorized-signature.png", "document-assets/vendor-authorized-signature.png", 480)
    save_jpg("logo.png", "logo.jpg", 512, 85)
    save_jpg("document-assets/agreement-stamp-paper.png", "document-assets/agreement-stamp-paper.jpg", 800, 85)
